File: packages/backend/src/auth/sessions.ts
import { createHash, randomBytes } from 'node:crypto'
import type { ControlService } from '../storage/control'
import type { User } from './identity'

export interface WebSessionsOptions {
  /** How long a session lives from its last renewal. Default 30 days. */
  ttlMs?: number
  /** A request that finds less than this left renews the session to a full ttl. Default 15 days. */
  renewBelowMs?: number
  /** The clock — tests move it. */
  now?: () => number
}

const DAY_MS = 24 * 60 * 60 * 1000

/**
 * Cookie sessions (`product.md` § User system): a 256-bit random token in
 * the browser, its SHA-256 in `web_sessions`, sliding expiry. The token
 * itself is never stored.
 */
export class WebSessions {
  private readonly ttlMs: number
  private readonly renewBelowMs: number
  private readonly now: () => number

  constructor(
    private readonly control: ControlService,
    options: WebSessionsOptions = {},
  ) {
    this.ttlMs = options.ttlMs ?? 30 * DAY_MS
    this.renewBelowMs = options.renewBelowMs ?? 15 * DAY_MS
    this.now = options.now ?? Date.now
  }

  async open(userId: string): Promise<{ token: string; expiresAt: Date }> {
    const token = randomBytes(32).toString('base64url')
    const expiresAt = new Date(this.now() + this.ttlMs)
    await this.control.createWebSession({ tokenHash: hashSessionToken(token), userId, expiresAt: expiresAt.toISOString() })
    return { token, expiresAt }
  }

  /** The session's user, or null when the token is unknown or expired; renews a session near its end. */
  async resolve(token: string): Promise<{ user: User; expiresAt: Date; renewed: boolean } | null> {
    const tokenHash = hashSessionToken(token)
    const session = await this.control.getWebSession(tokenHash)
    if (!session) return null
    const now = this.now()
    let expiresAt = new Date(session.expiresAt)
    if (expiresAt.getTime() <= now) {
      await this.control.deleteWebSession(tokenHash)
      return null
    }
    const user = await this.control.getUser(session.userId)
    if (!user) return null
    let renewed = false
    if (expiresAt.getTime() - now < this.renewBelowMs) {
      expiresAt = new Date(now + this.ttlMs)
      await this.control.extendWebSession(tokenHash, expiresAt.toISOString())
      renewed = true
    }
    return { user, expiresAt, renewed }
  }

  async close(token: string): Promise<void> {
    await this.control.deleteWebSession(hashSessionToken(token))
  }
}

function hashSessionToken(token: string): string {
  return createHash('sha256').update(token).digest('hex')
}

File: crates/backend/src/auth/sessions.rs
//! Browser sessions (`backend.md` § Authentication and ownership): the
//! cookie holds a random 256-bit token, storage holds the token's SHA-256,
//! and a session expires 30 days after its last renewal.

use demi_core::Timestamp;
use demi_web_api::auth::UserDto;
use demi_web_api::ids::UserId;
use jiff::SignedDuration;
use sha2::{Digest, Sha256};

use crate::storage::StorageError;
use crate::storage::control::ControlService;

/// How long a session lives after its last renewal, and how little of that
/// must remain for a request to renew it.
#[derive(Debug, Clone, Copy)]
pub(crate) struct SessionPolicy {
    pub(crate) lifetime: SignedDuration,
    pub(crate) renew_below: SignedDuration,
}

const DAY_SECONDS: i64 = 24 * 60 * 60;

pub(crate) const SESSION_POLICY: SessionPolicy = SessionPolicy {
    lifetime: SignedDuration::from_secs(30 * DAY_SECONDS),
    renew_below: SignedDuration::from_secs(15 * DAY_SECONDS),
};

/// The SHA-256 of a session or device token in lowercase hexadecimal: what
/// storage keeps instead of the token (`storage.md` § Passwords and
/// credentials at rest).
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct TokenHash(String);

impl TokenHash {
    pub(crate) fn of(token: &str) -> Self {
        Self(hex::encode(Sha256::digest(token.as_bytes())))
    }

    pub(crate) fn as_str(&self) -> &str {
        &self.0
    }
}

/// A new session: the token for the cookie and when the session expires.
pub(crate) struct OpenedSession {
    pub(crate) token: String,
    pub(crate) expires_at: Timestamp,
}

/// A live session's user and expiry, and whether this request renewed it.
pub(crate) struct ResolvedSession {
    pub(crate) user: UserDto,
    pub(crate) expires_at: Timestamp,
    pub(crate) renewed: bool,
}

/// Opens, resolves and closes sessions in the control database.
pub(crate) struct WebSessions {
    control: ControlService,
}

impl WebSessions {
    pub(crate) fn new(control: ControlService) -> Self {
        Self { control }
    }

    pub(crate) async fn open(&self, user: UserId) -> Result<OpenedSession, StorageError> {
        let token = hex::encode(rand::random::<[u8; 32]>());
        let expires_at = self
            .control
            .open_web_session(TokenHash::of(&token), user, SESSION_POLICY)
            .await?;
        Ok(OpenedSession { token, expires_at })
    }

    /// The session a cookie's token names, or `None` when it names no live
    /// session.
    pub(crate) async fn resolve(&self, token: &str) -> Result<Option<ResolvedSession>, StorageError> {
        self.control
            .resolve_web_session(TokenHash::of(token), SESSION_POLICY)
            .await
    }

    pub(crate) async fn close(&self, token: &str) -> Result<(), StorageError> {
        self.control.close_web_session(TokenHash::of(token)).await
    }
}
